### PR TITLE
[MINOR][DOC] Add note regarding proper usage of QueryExecution.toRdd

### PR DESCRIPTION
**Pull request by HeartSaVioR**
## What changes were proposed in this pull request?

This proposes adding a note on `QueryExecution.toRdd` regarding Spark's internal optimization callers would need to indicate.

## How was this patch tested?

This patch is a documentation change.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
Retest this, please

**Comment by HeartSaVioR** [comment]
cc. @cloud-fan

**Comment by cloud-fan** [comment]
LGTM

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 1]
```diff
@@ -85,7 +85,14 @@ class QueryExecution(
...
 
-  /** Internal version of the RDD. Avoids copies and has no schema */
+  /**
+   * Internal version of the RDD. Avoids copies and has no schema.
+   * Note for callers: Spark may apply various optimization including reusing object: this means
+   * the row is valid only for the iteration it is retrieved. You should avoid storing row and
+   * accessing after iteration. (Calling `collect()` is one of known bad usage.)
+   * If you want to store these rows into collection, please apply some converter or copy row
+   * which produces new object per iteration.
+   */
```
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: @HeartSaVioR Should we point the users to dataset.rdd method where the conversion is already applied ?

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: Yeah that's good suggestion for end users (not Spark developers). Will add.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 2]
```diff
@@ -92,6 +92,8 @@ class QueryExecution(
    * accessing after iteration. (Calling `collect()` is one of known bad usage.)
    * If you want to store these rows into collection, please apply some converter or copy row
    * which produces new object per iteration.
+   * Given QueryExecution is not a public class, end users are discouraged to use this: please
+   * user `Dataset.rdd` instead which conversion will be applied.
```
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: user -> use
which -> in which
or
which -> where  ?

**Comment by dilipbiswal** [comment]
LGTM with a very minor comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala (hunk 2 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: Nice finding. Applied.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala (hunk 1 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: BTW, I don't think it's an API though .. technically we don't have to worry about end users.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Review by viirya — APPROVED** [review]

**Pull request closed without merging**